src/inference/involution_dsl.jl

```diff
@@ -0,0 +1,440 @@
+        next_output_index += 1
+    end
+
+    cont_u_back_key_to_index = Dict()
+    for (addr, v) in first_pass_state.u_cont_writes
+        cont_u_back_key_to_index[addr] = next_output_index
+        next_output_index += 1
+    end
+
+    function f_array(input_arr::AbstractArray{T}) where {T <: Real}
+
+        # closing over:
+        # - trace, u
+        # - u_key_to_index, t_key_to_index, cont_constraints_key_to_index, cont_u_back_key_to_index
+        # - proposal_args, proposal_retval
+
+        n_output = length(cont_constraints_key_to_index) + length(cont_u_back_key_to_index)
+        output_arr = Vector{T}(undef, n_output)
+
+        jacobian_pass_state = JacobianPassState(
+            trace, u, input_arr, output_arr, 
+            t_key_to_index, u_key_to_index,
+            cont_constraints_key_to_index,
+            cont_u_back_key_to_index)
+
+        # mutates the state
+        involution.fn!(
+            jacobian_pass_state, get_args(trace), proposal_args, proposal_retval)
+
+        # return the output array
+        output_arr
+    end
+
+    # TODO XXX proposal_retval cannot depend on continuous random choices in model
+    # or proposal (we might be abel to relax this later using choice_gradients) XXX
+
+    (first_pass_state.constraints, first_pass_state.u_back,
+     first_pass_state.marked_as_retained,
+     t_key_to_index, input_arr, f_array)
+end
+
+# callable
+
+function (involution::InvolutionDSLProgram)(trace, u::ChoiceMap, proposal_retval, proposal_args; check=false)
+
+    # apply the involution, and get metadata back about it, including the
+    # function f_array from arrays to arrays, that is the restriction of the
+    # involution to just a subset of the relevant continuous random choices
+    (constraints, u_back, marked_as_retained, t_key_to_index, input_arr, f_array) = apply_involution(
+        involution, trace, u, proposal_args, proposal_retval)
+
+    # update model trace
+    (new_trace, model_weight, _, discard) = update(
+        trace, get_args(trace), map((_) -> NoChange(), get_args(trace)), constraints)
+
+    # check the user's retained assertions
+    if check
+        for addr in marked_as_retained
+            has_value(discard, addr) && error("Address $addr was marked as retained, but was not")
+        end
+    end
+
+    # compute Jacobian matrix of f_array, where columns are inputs, rows are outputs
+    J = ForwardDiff.jacobian(f_array, input_arr)
+    @assert size(J)[2] == length(input_arr)
+    num_outputs = size(J)[1]
+    #println("num_outputs: $num_outputs")
+
+    # remove columns for inputs from the trace that were retained
+    # NOTE: these columns did not have to be computed in the first place, if
+    # the user had marked the associated random choices as retained
+    keep = fill(true, length(input_arr))
+    for (addr, index) in t_key_to_index
+        if !has_value(discard, addr)
+            keep[index] = false
+        end
+    end
+    J = J[:,keep]
```

Comment:
I think I see what's going on here. 

So: if the user writes `read` but doesn't make the corresponding `write` for a continuous choice they want "implicitly copied," that's fine -- we can just detect and remove the relevant rows from the Jacobian before computing the determinant. 

But see above re: whether we might call `update` from within `apply_involution`, and remove the need for the user to do this manually (though again, I could totally be missing something!).

Reply:
This code is now removed